Inst tokens

nils-herrmann opened the pull request:
Changes:
- Insttokens can be passed as a list of tuples: `[('key1', 'token1'), ('key2', 'token2')]`
- Insttokens are saved as a string in the config: `key1: token1, key2: token2`
- The new strategy to get content uses first the tokens and then the keys.

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -111,13 +111,14 @@ def get_content(url, api, params=None, **kwds):
+
     else:
         header = key_header
+        resp = session.get(url, headers=header, params=params, timeout=timeout, proxies=proxies)
```
I don't think that's necessary. If `token_key=True`, you can simply set proxies to None or an empty dict.

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -126,8 +127,7 @@ def get_content(url, api, params=None, **kwds):
             token_header['X-ELS-Insttoken'] = token
-            resp = session.get(url, headers=token_header, proxies=proxies,
-                               params=params, timeout=timeout)
+            resp = session.get(url, headers=token_header, params=params, timeout=timeout)
```
It's important that this query is the same as above.

Michael-E-Rose commented on pybliometrics/utils/create_config.py:
```diff
@@ -5,8 +5,8 @@
 def create_config(config_dir: Optional[Union[str, Path]] = None,
-                  keys: Optional[list[str]] = None,
-                  insttoken: Optional[str] = None
+                  custom_keys: Optional[list[str]] = None,
+                  custom_tokens: Optional[list[tuple[str, str]]] = None
```
For backward compatibility, we must not change the parameter names!

nils-herrmann replied:
True!

Michael-E-Rose commented on pybliometrics/utils/create_config.py:
```diff
@@ -33,23 +33,36 @@ def create_config(config_dir: Optional[Union[str, Path]] = None,
+    new_keys, new_tokens = None, None
+    if custom_keys:
+        if not isinstance(custom_keys, list):
             raise ValueError("Parameter `keys` must be a list.")
-        key = ", ".join(keys)
-        token = insttoken
-    else:
+        new_keys = ", ".join(custom_keys)
+    if custom_tokens:
+        if not isinstance(custom_tokens, list):
+            raise ValueError("Parameter `inst_tokens` must be a list of tuples. "\
+                             "E.g. [('key_1', 'token_1'), ('key_2', 'token_2')]")
+        new_tokens = ", ".join([f"{key}:{value}" for key, value in custom_tokens])
```
I think an easier setup is simply to zip the keys and the tokens. For robustness, use the first `n = len(tokens)` keys. That means, `insttoken` can remain to be a simple list.

nils-herrmann replied:
Good idea! Much simpler

Michael-E-Rose commented on pybliometrics/utils/create_config.py:
```diff
@@ -5,8 +5,8 @@
-                  keys: Optional[list[str]] = None,
-                  insttoken: Optional[str] = None
+                  custom_keys: Optional[list[str]] = None,
+                  custom_tokens: Optional[list[tuple[str, str]]] = None
```
`insttoken` should be a list, I agree. But not a list of tuples. That will be too complex.

Michael-E-Rose commented on pybliometrics/utils/startup.py:
```diff
@@ -8,11 +8,14 @@
-def init(config_dir: Optional[str] = CONFIG_FILE, keys: Optional[list[str]] = None) -> None:
+def init(config_dir: Optional[str] = CONFIG_FILE,
+         keys: Optional[list[str]] = None,
+         inst_tokens: Optional[list[tuple[str, str]]] = None) -> None:
```
Try to keep the `inst_token` a simple list.

Michael-E-Rose commented on pybliometrics/utils/startup.py:
```diff
@@ -79,10 +97,36 @@ def get_config() -> Type[ConfigParser]:
     return keys
+
+
+def _token_text_to_tuple(raw_token_text: str) -> list[tuple[str, str]]:
```
Too complex. I should be possible to do without. Have a look a zip().

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -99,19 +108,36 @@ def get_content(url, api, params=None, **kwds):
-            header['X-ELS-APIKey'] = keys[0].strip()
-            resp = session.get(url, headers=header, proxies=proxies,
-                               params=params, timeout=timeout)
+            key_header['X-ELS-APIKey'] = keys[0].strip()
+            resp = session.get(url, headers=key_header, proxies=proxies,
```
Why renaming?

nils-herrmann replied:
What do you mean by renaming ?

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -99,19 +108,36 @@ def get_content(url, api, params=None, **kwds):
+    # If 429 try other tokens
+    while resp.status_code == 429:
+        try:
+            insttokens.pop(0)
+            shuffle(insttokens)
+            key, token = insttokens[0]
```
My suggestion would be to create a list of tuples combining keys and insttokens (as many keys are needed, assuming that the first key belongs to the first insttoken). Then you simply shuffle that list.

Michael-E-Rose reviewed (changes requested):
Try to simplify by zipping keys and insttoken in the specified order. And do not rename parameters of public functions.

Michael-E-Rose commented on pybliometrics/utils/create_config.py:
```diff
@@ -33,23 +33,36 @@ def create_config(config_dir: Optional[Union[str, Path]] = None,
     config.add_section('Authentication')
+
+    # Get keys and tokens
+    new_keys, new_tokens = None, None
```
Why are these two objects necessary? Why not simply use `keys` and 'tokens'?

nils-herrmann replied:
Let's simply use `keys` and `tokens`. I just wanted to avoid having one variable with several data types (`str` and `list[str]`) in the code.

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -67,30 +67,39 @@ def get_content(url, api, params=None, **kwds):
+    token_key, insttoken = None, None
     if "insttoken" in params:
-        header['X-ELS-Insttoken'] = params.pop("insttoken")
+        token_key, insttoken = params.pop("insttoken")
```
Why does the key come from `params["insttoken"]`? The key must be passed through `params["apikey"]`.

nils-herrmann replied:
True

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -99,19 +108,36 @@ def get_content(url, api, params=None, **kwds):
-                       timeout=timeout)
+    # Either use token header or key header
+    if token_key:
+        header = token_header
```
What is this good for?

nils-herrmann replied:
There is actually no need to define a new `header` variable. We can directly pass the `token_header` and `key_header` to the `session.get()`

Michael-E-Rose commented on pybliometrics/utils/startup.py:
```diff
@@ -61,15 +69,43 @@ def check_default_paths(config: Type[ConfigParser], config_path: Path) -> None:
+    keys_tokens_diff = len(keys) - len(tokens)
+
+    if no_keys_no_tokens:
+        raise ValueError('No API keys or InstTokens found. '
+                         'Please provide at least one key or token. '
+                         'For more information visit: '
+                         'https://pybliometrics.readthedocs.io/en/stable/configuration.html')
+    elif tokens_no_keys:
+        raise ValueError('API tokens found but not corresponding keys. '
+                             'Please provide the keys that correspond to the tokens. '
+                             'For more information visit: '
+                             'https://pybliometrics.readthedocs.io/en/stable/configuration.html')
+    elif keys_and_tokens:
+        if keys_tokens_diff < 0:
+            raise ValueError('More tokens than keys found. '
+                             'Please provide all the keys that correspond to the tokens. '
+                             'For more information visit: '
+                             'https://pybliometrics.readthedocs.io/en/stable/configuration.html')
```
Stay consistent with the words please. Call it "API keys" and "InstTokens", and don't deviate from this.

Michael-E-Rose reviewed (changes requested):
I don't get why separate `token_header` and `inst_header` are necessary.

As for the edge cases:
- We can accept that there are more keys than tokens. In this case we will simply use as many keys as there are tokens, counting from the top. The rest is discarded. No need to raise an error for that. That is important because you will have more normal keys than keys with InstTokens. But depending on the network, I need the InstTokensor not. When I take my computer home, I need the InstTokens, in which case I have them in my configuration file. When I take my computer to work, I do not need the InstTokens, in which case I comment out the InstTokens. But we must not force the users to comment out all keys when they get home.
- We cannot accept more InstTokens than keys, so that is good when it raises an error.

Ideally, the zipping of the keys and tokens happens in `get_content()`, for reasons of transparency.

nils-herrmann commented:
- We can use one header. We just have to delete `header['X-ELS-Insttoken']` before doing queries with keys.
- No error is raised for the case of more `keys` than `tokens`
- Using InstToken-**keys** without an InstToken results in a `Scopus401Error` (`Invalid API Key`). I added the `401` error as a condition to loop over other keys and tokens. This enables users to comment out the `InstToken` in the config.
- I moved the zipping to `get_content()`

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -6,7 +6,7 @@
 from pybliometrics import __version__
 from pybliometrics import exception
-from pybliometrics.utils.startup import get_config, get_keys, _throttling_params
+from pybliometrics.utils.startup import get_config, get_keys, get_insttokens, _throttling_params
```
Sort alphabetically

Michael-E-Rose commented on pybliometrics/utils/create_config.py:
```diff
@@ -33,23 +33,35 @@ def create_config(config_dir: Optional[Union[str, Path]] = None,
+        insttoken = input(prompt_token)
+
+    # Set keys and tokens in config
+    if keys:
```
That is unnecessary because a key needs to be present to continue

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -67,30 +67,43 @@ def get_content(url, api, params=None, **kwds):
+                   'User-Agent': user_agent}
+
+    # Get tokens and create header
+    token_key, insttoken = None, None
```
Seems unnecessary because it will always be redefined

nils-herrmann replied:
In case there are no `insttokens` found both variables won't be redefined. We need to set them to `None` to avoid an error in [line 116](https://github.com/nils-herrmann/pybliometrics/blob/705736494b4c658ebe230d4429f8b63ea94d0dfc/pybliometrics/utils/get_content.py#L116).

Michael-E-Rose replied:
oh, now I see, there is an `elif`!
Anyways, if you could simplify this, maybe as if - elif - else statement, that would be great. This function involves a great deal of variables and conditions, every reduction in complexity is a win.

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -67,30 +67,43 @@ def get_content(url, api, params=None, **kwds):
+        token_key, insttoken = insttokens[0]
+
+    # Get keys and create header
+    key = None
```
Seems unnecessary because it will always be redefined

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -67,30 +67,43 @@ def get_content(url, api, params=None, **kwds):
-    if "apikey" in params:
-        header['X-ELS-APIKey'] = params.pop("apikey")
+    # Base header
+    base_header = {'Accept': 'application/json',
+                   'User-Agent': user_agent}
```
if we move this further down, we can add the key, which always needs to be present.

Michael-E-Rose commented on pybliometrics/utils/get_content.py:
```diff
@@ -99,19 +112,44 @@ def get_content(url, api, params=None, **kwds):
+        header = {**base_header,
+                  'X-ELS-APIKey': token_key,
+                  'X-ELS-Insttoken': insttoken}
+        resp = session.get(url, headers=header, params=params, timeout=timeout)
+    else:
+        header = {**base_header,
+                  'X-ELS-APIKey': key}
+        resp = session.get(url, headers=header, params=params, timeout=timeout, proxies=proxies)
```
So proxies may not be used when InstTokens are in place. Can't we just enforce an empty dict and use a common `resp = ...` line of code?

nils-herrmann replied:
The problem is that if we get a 429 or 401 we have to set the proxies again to the ones of the config. I think it is more transparent if we have two `session.get(...)`.

Michael-E-Rose commented on pybliometrics/utils/startup.py:
```diff
@@ -61,6 +69,34 @@ def check_default_paths(config: Type[ConfigParser], config_path: Path) -> None:
+    keys = get_keys()
+    insttokens = get_insttokens()
+    # 3 problematic cases
+    no_keys_no_insttokens = True if not keys and not insttokens else False
+    insttokens_no_keys = True if insttokens and not keys else False
+    keys_and_insttokens = True if keys and insttokens else False
```
Can be simplified: `no_keys_no_insttokens = not keys and not insttokens`, etc.

Michael-E-Rose commented on pybliometrics/utils/startup.py:
```diff
@@ -69,7 +105,7 @@ def create_cache_folders(config: Type[ConfigParser]) -> None:
             view_path.mkdir(parents=True, exist_ok=True)
 
 
-def get_config() -> Type[ConfigParser]:
```
We need to keep the `Type[]` here because of Python 3.9

Michael-E-Rose reviewed (changes requested):
Now I just request a few cosmetically relevant changes
